Documentation/design-docs/struct-abi.md

```diff
@@ -0,0 +1,120 @@
+=============================
+Problem Statement
+-----------------
+The current implementation of ABI (Application Binary Interface, aka calling
+convention) support in RyuJIT is problematic in a number of areas, especially
+when it comes to the handling of structs (aka value types).
+
+- RyuJIT currently supports 4 target architectures: x86, x64 (aka x86-64), ARM
+  and ARM64, with two different ABIs for x64 (Windows and Linux).
+  These each have unique requirements, yet these requirements are expressed in
+  the code programmatically, with #ifdefs, and yet even where the requirements
+  are shared, they are often handled in different code paths.
+
+- When passing or returning structs, the code generator sometimes requires 
+  that the struct must be copied to or from memory. The morpher (`fgMorphArgs()`)
+  attempts to discern these cases, and create copies when necessary, but sometimes it
+  makes copies when they aren't needed.
+
+- Even in cases where the code generator currently requires the struct to be
+  in memory, it could be enhanced to handle the in-register case:
+  - Currently, when we have a register-passed struct that fits in a register,
+    but that doesn't have a single field of a matching type,
+    `fgMorphArgs()` generates a `GT_LCL_FLD` of the appropriate scalar type
+    to reference the value. This forces the struct to be marked `lvDoNotEnregister`.
+    However, the backend has support for performing the necessary move in
+    some cases (e.g. when a struct with a single field of `TYP_DOUBLE` is passed
+    in an integer register as `TYP_LONG`), by generating a `GT_BITCAST` to move
+    the value to the appropriate register.
+  - In other cases (e.g. a struct with two `TYP_INT` fields in registers), the
+    backend should be able to generate the necessary code to place the fields
+    in the necessary register(s).
+
+- Even when the requirements are similar, the IL representation, as well as the
+  transformations performed by `fgMorphArgs()`, are not the same.
+
+- Much of the information about each argument is contained in the `fgArgInfo`
+  on the `GT_CALL` node. It in turn contains an `argTable` with an entry for
+  each argument. However, this information is not complete, especially on
+  x64/Linux where repeated calls are made to the VM to obtain the struct
+  descriptor.
+
+- The functionality of `fgMorphArgs()` combines the determination of the ABI
+  requirements, which sets up the `fgArgInfo` and `argTable`, with the IR
+  transformations required to ensure that the arguments of the `GT_CALL` are
+  in the appropriate form.
+
+- When `fgCanFastTailCall()` is called, it doesn't yet have the `fgArgInfo`,
+  so it must duplicate some of the analysis that is done in `fgMorphArgs()`
+
+High-Level Proposed Design
+--------------------------
+This is a preliminary design, and is likely to change as the implementation proceeds:
+
+First, the `fgArgInfo` is extended to contain all the information needed to determine
```

Comment:
This would also, in theory, make it easier to support new calling conventions (like `__vectorcall` on Windows), correct?

Reply:
Right! I was just thinking about that yesterday during a discussion on the topic of HW intrinsics.